File: frontend/src/types/Player.ts
export interface Team {
  id: number;
  name: string;
  city: string;
}

export interface Player {
  id: number;
  name: string;
  position: string;
  nationality: string;
  jersey_number: number;
  birth_date: string;
  height: string;
  weight: number;
  handedness: string;
  active_status: boolean;

  // Regular season statistics
  regular_season_goals: number;
  regular_season_assists: number;
  regular_season_points: number;
  regular_season_games_played: number;

  // Playoff statistics
  playoff_goals: number;
  playoff_assists: number;
  playoff_points: number;
  playoff_games_played: number;
  // Combined statistics
  games_played: number;
  goals: number;
  assists: number;
  points: number;

  team: Team;
}

export interface PlayersResponse {
  players: Player[];
  count: number;
}

export type SearchField =
  | 'all'
  | 'name'
  | 'position'
  | 'team'
  | 'nationality'
  | 'jersey_number';

export interface SearchConfig {
  value: SearchField;
  label: string;
  searchKey?: keyof Player | 'team.name';
}

export const SEARCHABLE_FIELDS: SearchConfig[] = [
  { value: 'all', label: 'All Fields' },
  { value: 'name', label: 'Name', searchKey: 'name' },
  { value: 'position', label: 'Position', searchKey: 'position' },
  { value: 'team', label: 'Team', searchKey: 'team.name' },
  { value: 'nationality', label: 'Nationality', searchKey: 'nationality' },
  {
    value: 'jersey_number',
    label: 'Jersey Number',
    searchKey: 'jersey_number',
  },
];

// Sort field type - matches backend SortFieldType
export type SortField =
  | 'name'
  | 'position'
  | 'team'
  | 'jersey_number'
  | 'active_status'
  | 'regular_season_goals'
  | 'regular_season_assists'
  | 'regular_season_points'
  | 'regular_season_games_played'
  | 'playoff_goals'
  | 'playoff_assists'
  | 'playoff_points'
  | 'playoff_games_played'
  | 'games_played'
  | 'goals'
  | 'assists'
  | 'points';

export type SortDirection = 'asc' | 'desc';

export interface SortConfig {
  field: SortField;
  label: string;
  displayName: string;
}

export const SORTABLE_FIELDS: SortConfig[] = [
  { field: 'name', label: 'Name', displayName: 'Name' },
  { field: 'position', label: 'Position', displayName: 'Position' },
  { field: 'team', label: 'Team', displayName: 'Team' },
  { field: 'jersey_number', label: 'Jersey #', displayName: 'Jersey #' },
  { field: 'active_status', label: 'Status', displayName: 'Status' },
  {
    field: 'regular_season_goals',
    label: 'RS Goals',
    displayName: 'Regular Season Goals',
  },
  {
    field: 'regular_season_assists',
    label: 'RS Assists',
    displayName: 'Regular Season Assists',
  },
  {
    field: 'regular_season_points',
    label: 'RS Points',
    displayName: 'Regular Season Points',
  },
  {
    field: 'regular_season_games_played',
    label: 'RS Games',
    displayName: 'Regular Season Games',
  },
  { field: 'playoff_goals', label: 'PO Goals', displayName: 'Playoff Goals' },
  {
    field: 'playoff_assists',
    label: 'PO Assists',
    displayName: 'Playoff Assists',
  },
  {
    field: 'playoff_points',
    label: 'PO Points',
    displayName: 'Playoff Points',
  },
  {
    field: 'playoff_games_played',
    label: 'PO Games',
    displayName: 'Playoff Games',
  },
  {
    field: 'games_played',
    label: 'Games Played (Total)',
    displayName: 'Games Played (Total)',
  },
  {
    field: 'goals',
    label: 'Goals (Total)',
    displayName: 'Goals (Total)',
  },
  {
    field: 'assists',
    label: 'Assists (Total)',
    displayName: 'Assists (Total)',
  },
  { field: 'points', label: 'Points (Total)', displayName: 'Points (Total)' },
];

// Filter types - matches backend FilterFieldType
export type FilterField =
  | 'position'
  | 'team'
  | 'jersey_number'
  | 'active_status'
  | 'regular_season_goals'
  | 'regular_season_assists'
  | 'regular_season_points'
  | 'regular_season_games_played'
  | 'playoff_goals'
  | 'playoff_assists'
  | 'playoff_points'
  | 'playoff_games_played'
  | 'games_played'
  | 'goals'
  | 'assists'
  | 'points';

// Filter operator types
export type StringFilterOperator = '=' | '!=' | 'contains' | 'not_contains';
export type NumericFilterOperator = '=' | '!=' | '>' | '<' | '>=' | '<=';
export type BooleanFilterOperator = '=' | '!=';
export type FilterOperator =
  | StringFilterOperator
  | NumericFilterOperator
  | BooleanFilterOperator;

// Filter data type classification
export type FilterDataType = 'string' | 'numeric' | 'boolean';

export interface FilterConfig {
  field: FilterField;
  label: string;
  displayName: string;
  dataType: FilterDataType;
  operators: FilterOperator[];
}

export const FILTERABLE_FIELDS: FilterConfig[] = [
  {
    field: 'position',
    label: 'Position',
    displayName: 'Position',
    dataType: 'string',
    operators: ['=', '!=', 'contains', 'not_contains'],
  },
  {
    field: 'team',
    label: 'Team',
    displayName: 'Team',
    dataType: 'string',
    operators: ['=', '!=', 'contains', 'not_contains'],
  },
  {
    field: 'jersey_number',
    label: 'Jersey #',
    displayName: 'Jersey #',
    dataType: 'numeric',
    operators: ['=', '!=', '>', '<', '>=', '<='],
  },
  {
    field: 'goals',
    label: 'Goals',
    displayName: 'Goals',
    dataType: 'numeric',
    operators: ['=', '!=', '>', '<', '>=', '<='],
  },
  {
    field: 'assists',
    label: 'Assists',
    displayName: 'Assists',
    dataType: 'numeric',
    operators: ['=', '!=', '>', '<', '>=', '<='],
  },
  {
    field: 'points',
    label: 'Points',
    displayName: 'Points',
    dataType: 'numeric',
    operators: ['=', '!=', '>', '<', '>=', '<='],
  },
  {
    field: 'active_status',
    label: 'Status',
    displayName: 'Status',
    dataType: 'boolean',
    operators: ['=', '!='],
  },
];

// Filter interface
export interface PlayerFilter {
  field: FilterField;
  operator: FilterOperator;
  value: string | number | boolean;
}

// Operator display labels
export const OPERATOR_LABELS: Record<FilterOperator, string> = {
  '=': 'equals',
  '!=': 'does not equal',
  contains: 'contains',
  not_contains: 'does not contain',
  '>': 'greater than',
  '<': 'less than',
  '>=': 'greater than or equal to',
  '<=': 'less than or equal to',
};

// Boolean value options for active_status
export const BOOLEAN_OPTIONS = [
  { value: true, label: 'Active' },
  { value: false, label: 'Retired' },
];

// Extended API response including sorting and filtering info
export interface PlayersApiResponse {
  players: Player[];
  count: number;
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  search_query: string | null;
  search_field: SearchField;
  sort_by: SortField;
  sort_order: SortDirection;
  filters: PlayerFilter[];
}

File: backend/app/schemas/player.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, Field, validator

# Search field options - matches frontend SearchField type
SearchFieldType = Literal[
    "all", "name", "position", "team", "nationality", "jersey_number"
]

# Sort field options - matches database fields and team join
SortFieldType = Literal[
    "name",
    "position", 
    "team",
    "jersey_number",
    "active_status",
    "regular_season_games_played",
    "regular_season_goals",
    "regular_season_assists",
    "regular_season_points", 
    "playoff_games_played",
    "playoff_goals",
    "playoff_assists",
    "playoff_points",
    "games_played",
    "goals", 
    "assists", 
    "points", 
]

# Sort direction options
SortDirectionType = Literal["asc", "desc"]

# Filter field options - fields that can be filtered
FilterFieldType = Literal[
    "position", 
    "team", 
    "jersey_number", 
    "active_status",
    "regular_season_games_played",
    "regular_season_goals",
    "regular_season_assists", 
    "regular_season_points",
    "playoff_games_played",
    "playoff_goals",
    "playoff_assists",
    "playoff_points", 
    "games_played",
    "goals", 
    "assists", 
    "points", 
]

# Filter operators for different data types
StringFilterOperator = Literal["=", "!=", "contains", "not_contains"]
NumericFilterOperator = Literal["=", "!=", ">", "<", ">=", "<="]
BooleanFilterOperator = Literal["=", "!="]

FilterOperatorType = (
    StringFilterOperator | NumericFilterOperator | BooleanFilterOperator
)


class PlayerFilter(BaseModel):
    """
    Individual filter configuration.
    """

    field: FilterFieldType = Field(..., description="Field to filter on")
    operator: FilterOperatorType = Field(..., description="Filter operator")
    value: str | int | bool = Field(..., description="Filter value")

    @validator("operator")
    def validate_operator_for_field(cls, operator, values):
        """
        Validate that the operator is appropriate for the field type.
        """
        field = values.get("field")

        # String fields: position, team
        if field in ["position", "team"]:
            if operator not in ["=", "!=", "contains", "not_contains"]:
                raise ValueError(
                    f"Invalid operator '{operator}' for string field '{field}'"
                )

        # Numeric fields: jersey_number, all stat fields
        elif field in [
            "jersey_number",
            "regular_season_games_played",
            "regular_season_goals",
            "regular_season_assists", 
            "regular_season_points",
            "playoff_games_played",
            "playoff_goals",
            "playoff_assists",
            "playoff_points",
            "games_played",
            "goals", 
            "assists", 
            "points",
        ]:
            if operator not in ["=", "!=", ">", "<", ">=", "<="]:
                raise ValueError(
                    f"Invalid operator '{operator}' for numeric field '{field}'"
                )

        # Boolean field: active_status
        elif field == "active_status":
            if operator not in ["=", "!="]:
                raise ValueError(
                    f"Invalid operator '{operator}' for boolean field '{field}'"
                )

        return operator


class PlayerSearchParams(BaseModel):
    """
    Query parameters for player search endpoint.
    """

    search: str | None = Field(None, description="Search query string")
    field: SearchFieldType = Field("all", description="Field to search in")
    page: int = Field(1, ge=1, description="Page number (starts from 1)")
    limit: int = Field(20, ge=1, le=100, description="Number of results per page")
    sort_by: SortFieldType = Field("name", description="Field to sort by")
    sort_order: SortDirectionType = Field(
        "asc", description="Sort direction (asc or desc)"
    )
    filters: list[PlayerFilter] = Field([], description="List of filters to apply")


class TeamResponse(BaseModel):
    """
    Team information in API responses.
    """

    id: int
    name: str
    city: str


class PlayerResponse(BaseModel):
    """
    Player information in API responses.
    """

    id: int
    name: str
    position: str
    nationality: str
    jersey_number: int
    birth_date: str
    height: str
    weight: int
    handedness: str
    active_status: bool
    
    # Regular season statistics
    regular_season_goals: int
    regular_season_assists: int
    regular_season_points: int
    regular_season_games_played: int
    
    # Playoff statistics
    playoff_goals: int
    playoff_assists: int
    playoff_points: int
    playoff_games_played: int

    # Combined statistics
    games_played: int
    goals: int
    assists: int
    points: int
    
    team: TeamResponse


class PlayersListResponse(BaseModel):
    """
    Response model for players list endpoint.
    """

    players: list[PlayerResponse]
    count: int
    total: int
    page: int
    limit: int
    total_pages: int


class PlayerSearchResponse(BaseModel):
    """
    Response model for player search results.
    """

    players: list[PlayerResponse]
    count: int
    total: int
    page: int
    limit: int
    total_pages: int
    search_query: str | None
    search_field: SearchFieldType
    sort_by: SortFieldType
    sort_order: SortDirectionType
    filters: list[PlayerFilter]


class PlayerCreate(BaseModel):
    """
    Schema for creating a new player.
    """

    name: str = Field(..., min_length=1, max_length=100, description="Player name")
    jersey_number: int = Field(..., ge=0, le=99, description="Jersey number (0-99)")
    position: str = Field(..., description="Playing position")
    team_id: int = Field(..., gt=0, description="Team ID")
    nationality: str = Field(..., min_length=1, description="Player nationality")
    birth_date: date = Field(..., description="Date of birth")
    height: str = Field(..., description="Player height (e.g., 6'2\")")
    weight: int = Field(..., gt=0, description="Player weight in lbs")
    handedness: str = Field(..., description="Left or Right")
    active_status: bool = Field(True, description="Whether player is currently active")

    # Regular season statistics with defaults
    regular_season_games_played: int = Field(0, ge=0, description="Regular season games played")
    regular_season_goals: int = Field(0, ge=0, description="Regular season goals")
    regular_season_assists: int = Field(0, ge=0, description="Regular season assists")

    # Playoff statistics with defaults
    playoff_games_played: int = Field(0, ge=0, description="Playoff games played")
    playoff_goals: int = Field(0, ge=0, description="Playoff goals")
    playoff_assists: int = Field(0, ge=0, description="Playoff assists")

    @validator("position")
    def validate_position(cls, v):
        """Validate position is one of the allowed values."""
        allowed_positions = ["C", "LW", "RW", "D", "G"]
        if v not in allowed_positions:
            raise ValueError(f"Position must be one of {allowed_positions}")
        return v

    @validator("handedness")
    def validate_handedness(cls, v):
        """Validate handedness is Left or Right."""
        if v not in ["L", "R"]:
            raise ValueError("Handedness must be 'L' or 'R'")
        return v


class PlayerUpdate(BaseModel):
    """
    Schema for updating an existing player.
    """

    name: str = Field(..., min_length=1, max_length=100, description="Player name")
    jersey_number: int = Field(..., ge=0, le=99, description="Jersey number (0-99)")
    position: str = Field(..., description="Playing position")
    team_id: int = Field(..., gt=0, description="Team ID")
    nationality: str = Field(..., min_length=1, description="Player nationality")
    birth_date: date = Field(..., description="Date of birth")
    height: str = Field(..., description="Player height (e.g., 6'2\")")
    weight: int = Field(..., gt=0, description="Player weight in lbs")
    handedness: str = Field(..., description="Left or Right")
    active_status: bool = Field(..., description="Whether player is currently active")

    # Regular season statistics
    regular_season_games_played: int = Field(..., ge=0, description="Regular season games played")
    regular_season_goals: int = Field(..., ge=0, description="Regular season goals")
    regular_season_assists: int = Field(..., ge=0, description="Regular season assists")

    # Playoff statistics
    playoff_games_played: int = Field(..., ge=0, description="Playoff games played")
    playoff_goals: int = Field(..., ge=0, description="Playoff goals")
    playoff_assists: int = Field(..., ge=0, description="Playoff assists")

    @validator("position")
    def validate_position(cls, v):
        """Validate position is one of the allowed values."""
        allowed_positions = ["C", "LW", "RW", "D", "G"]
        if v not in allowed_positions:
            raise ValueError(f"Position must be one of {allowed_positions}")
        return v

    @validator("handedness")
    def validate_handedness(cls, v):
        """Validate handedness is Left or Right."""
        if v not in ["L", "R"]:
            raise ValueError("Handedness must be 'L' or 'R'")
        return v